docs/RFCS/20191028_easier_hash_sharded_indexes.md

```diff
@@ -0,0 +1,220 @@
+workloads by creating an index on a computed shard column. However, this feature still
+requires some relatively unattractive syntax to manually add a computed column which will
+act as the shard key. This is illustrated in the following example from issue [#39340]
+(https://github.com/cockroachdb/cockroach/issues/39340).
+
+Imagine we have an IOT application where we are tracking a bunch of devices and each
+device creates events. Sometimes we want to know which devices published events in some
+time period. We might start with the following schema.
+
+```sql
+CREATE TABLE events (
+    device_id
+        UUID,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (device_id, ts, event_id),
+    INDEX (ts)
+);
+```
+
+This schema would have a hot spot on that INDEX (ts) which would be rather unfortunate. We
+can alleviate this hot spot by sharding this time ordered index.
+
+```sql
+CREATE TABLE events (
+    device_id
+        UUID,
+    shard
+        INT8
+        AS (fnv32(device_id) % 8) STORED
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)),
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (device_id, ts, event_id),
+    INDEX (shard, ts)
+);
+```
+
+This isn't too big of a lift here because the device ID is easy to hash in sql. Imagine
+instead we had a primary key based on some other features:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts)
+);
+```
+
+In order to shard this we'll need something like:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    shard
+        INT8
+        AS (
+            fnv32(
+                concat(hex(product_id)),
+                owner::STRING,
+                serial_number
+            )
+            % 8
+        ) STORED
+        CHECK (shard IN (0, 1, 2, 3, 4, 5, 6, 7)),
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (
+        product_id,
+        owner,
+        serial_number,
+        ts,
+        event_id
+    ),
+    INDEX (shard, ts)
+);
+```
+
+We can see that this is starting to get heavy. The proposal is that we shoulder the burden
+of hashing and installing a check constraint behind a new syntax:
+
+```sql
+CREATE TABLE events (
+    product_id
+        INT8,
+    owner
+        UUID,
+    serial_number
+        VARCHAR,
+    event_id
+        UUID,
+    ts
+        TIMESTAMP,
+    data
+        JSONB,
+    PRIMARY KEY (product_id, owner, serial_number, ts, event_id),
+    INDEX (ts) USING HASH (product_id, owner, serial_number) WITH (BUCKET_COUNT=8)
+);
+```
+
+# Benchmarks 
```

Comment:
Other things to look at are performance for sharded secondary indexes which have sequential values and then also to look at range scan performance in both the primary and secondary index cases.